File: java/src/_0238.java
/*
 * @Author: whe
 * @Date: 2022-02-21 16:59:57
 */

public class _0238 {
    public int[] productExceptSelf(int[] nums) {
        int n = nums.length;
        int[] ans = new int[n];
        ans[0] = 1;
        for (int i = 1; i < n; i++) {
            ans[i] = ans[i - 1] * nums[i - 1];
        }
        int t = 1;
        for (int i = n - 2; i >= 0; i--) {
            ans[i] = nums[i + 1] * t  * ans[i];
            t = t * nums[i + 1];
        }
        return ans;
    }
}

File: go/solution/0238.go
/*
 * @Author: whe
 * @Date: 2022-02-21 16:47:12
 */

package solution

func productExceptSelf(nums []int) []int {
	ans := make([]int, len(nums))
	ans[0] = 1
	//ans[i]存对角线左边的乘积
	for i := 1; i < len(nums); i++ {
		ans[i] = ans[i-1] * nums[i-1]
	}
	//t存对角线右边乘积
	t := 1
	for i := len(nums) - 2; i >= 0; i-- {
		t = t * nums[i+1]
		ans[i] = ans[i] * t
	}
	return ans
}
